gc: take range tombstones into account when GC

aliher1911 opened the pull request:
Only last to commits are relevant to this PR, preceding ones are coming from uncommitted branch used for mvcc_bulk ops work.

Will update this comment once it becomes mergeable.

erikgrinaker committed:
storage: add experimental MVCC range key primitives

This patch adds initial experimental primitives for MVCC range keys,
which will be the foundation for MVCC range tombstones. They are based
on experimental Pebble range keys.

* Data structures:
  * `MVCCRangeKey`
  * internal `nil` value for range tombstones (as with point tombstones)

* `Engine` methods for mutating range keys:
  * `ExperimentalClearMVCCRangeKey()`
  * `ExperimentalClearAllMVCCRangeKeys()`
  * `ExperimentalPutMVCCRangeKey()`
  * `SupportsRangeKeys()`

* `SimpleMVCCIterator` methods for accessing range keys:
  * `HasPointAndRange()`
  * `RangeBounds()`
  * `RangeKeys()`

Range keys do not have a distinct identity, and should instead be
considered a key continuum: they will merge with abutting keys of the
same value, can be partially cleared, can split or merge along with
ranges, and so on. Bounded scans will truncate them to the scan bounds.

Range key support is implemented in `pebbleIterator` and
`intentInterleavingIter`, but not in the rest of the MVCC or KV APIs.
They are not persisted to disk either. Subsequent pull requests will
extend their functionality and integrate them with other components.

Release note: None

erikgrinaker committed:
batcheval: add range tombstone support for `DeleteRange`

This patch adds a parameter `UseExperimentalRangeTombstone` for
`DeleteRange`, which deletes the span using an MVCC range tombstone,
backed by an `ExperimentalDeleteRangeUsingTombstone` MVCC function. It
also adds an `ExperimentalMVCCRangeTombstones` version gate which must
be checked before using it.

This is an experimental implementation to allow writing range keys via
the KV API for testing and development purposes. It has significant
shortcomings, and will be fleshed out at a later time.

Release note: None

erikgrinaker committed:
storage: add `MVCCStats` for range keys

This patch adds `MVCCStats` tracking for range keys. Two new fields are
added to `MVCCStats`:

* `RangeKeyCount`: the number of (fragmented) range keys, not counting
  historical versions.

* `RangeKeyBytes`: the logical encoded byte size of all range keys.
  The latest version contributes the encoded key bounds, and all
  versions contribute encoded timestamps. Unlike point keys, which for
  historical reasons use a fixed-size timestamp contribution, this uses
  the actual variable-length timestamp size.

`ComputeStatsForRange()` has been extended to calculate the above
quantities, and additionally account for range tombstones themselves in
`GCBytesAge` along with their effect on point keys. All relevant call
sites have been updated to surface range keys for the MVCC iterators
passed to `ComputeStatsForRange()`.

Rudimentary range tombstone checks have been added during MVCC point
writes and intent resolution to correctly account for them in MVCC
statistics. Any further integration between point writes and range
tombstones (in particular, conflict handling) will be properly dealt
with later.

Range key stats are also adjusted during range splits and merges, which
will split and merge any range keys that straddle the split key. This
requires a single range key seek to the left and right of the split key
during these operations.

Release note: None

aliher1911 committed:
gc: take range tombstones into account when GC

Previously GC didn't take range tombstones into account, that prevented
values that are hidden by range tombstones from being collected.
This patch adds support for collection of versions hidden by range tombstones.

Release note: None

erikgrinaker committed:
storage: add stats tracing for MVCC history tests

This patch adds a `run` keyword `stats` which will compute and output
stats after a test. If `trace` is also enabled, it additionally outputs
the stats delta for each command.

Release note: None

erikgrinaker committed:
roachpb: add `Key.Prevish()` to find a previous key

This patch adds `Key.Prevish()`, which returns a previous key in
lexicographical sort order. This is needed to expand a latch span
leftwards to peek at any left-adjacent range keys.

It is impossible to find the exact immediate predecessor of a key,
because it can have an infinite number of `0xff` bytes at the end, so
this returns the nearest previous key right-padded with `0xff` up to the
given length. It is still possible for an infinite number of keys to
exist between `Key` and `Key.Prevish()`, as keys have unbounded length.

Release note: None

erikgrinaker committed:
storage: add `IterOptions.RangeKeyMaskingBelow`

This patch adds an iterator option `RangeKeyMaskingBelow`, which will
cause range keys at or below the given timestamp to mask any point
keys below them.

Release note: None

erikgrinaker committed:
storage: export some MVCC key encoding functions

Release note: None

erikgrinaker committed:
storage: re-introduce dual buffer for iterator bounds

This is necessary to avoid mutating the bounds in use by
`pebble.Iterator`, which can interfere with seek optimizations.

Release note: None

erikgrinaker committed:
storage: simplify `pebbleIterator` options comparisons

Since we now use dual buffers for bounds encoding, we can fully generate
the new `pebble.IterOptions` during `setOptions()` and compare them
directly with the existing options, which simplifies the code.

Release note: None

erikgrinaker committed:
storage: clear range keys in `Writer.Clear*Range` methods

This patch automatically clears any range keys in the `Writer` methods
`ClearRawRange`, `ClearMVCCRange`, and `ClearMVCCIteratorRange`.

Range keys are not cleared in `ClearMVCCVersions`, since this method is
specifically for clearing MVCC point key versions, and it is not
possible to clear range keys between versions of the same point key.

Release note: None

erikgrinaker committed:
storage: rename `Writer.Clear*Range` methods

This renames the `Clear*Range` methods on `Writer` in an attempt to
clarify their semantics, keeping their upcoming range key effects in
mind.

* `ClearMVCCRange` → `ClearMVCCVersions` (only point key versions)
* `ClearMVCCRangeAndIntents` → `ClearMVCCRange` (everything in span)
* `ClearIterRange` → `ClearMVCCIteratorRange` (uses `MVCCIterator`)

Release note: None

erikgrinaker committed:
batcheval: handle MVCC range tombstones in `ClearRange`

This patch makes `ClearRange` account for MVCC range tombstones when
updating MVCC stats.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81096)
<!-- Reviewable:end -->

aliher1911 committed:
gc: remove range tombstones during gc operations

Previously range tombstones were used in data deletion, but
wasn't removed from storage as part of GC process.
This patch introduces deletion of tombstones that are older
than gc threshold after all point values are deleted for the
range.

Release note: None
